dex/farm-with-locked-rewards/src/lib.rs

```diff
@@ -0,0 +1,195 @@
+elrond_wasm::derive_imports!();
+
+use common_structs::FarmTokenAttributes;
+use contexts::storage_cache::StorageCache;
+
+use farm::exit_penalty::{
+    DEFAULT_BURN_GAS_LIMIT, DEFAULT_MINUMUM_FARMING_EPOCHS, DEFAULT_PENALTY_PERCENT,
+};
+
+type EnterFarmResultType<BigUint> = EsdtTokenPayment<BigUint>;
+type CompoundRewardsResultType<BigUint> = EsdtTokenPayment<BigUint>;
+type ClaimRewardsResultType<BigUint> =
+    MultiValue2<EsdtTokenPayment<BigUint>, EsdtTokenPayment<BigUint>>;
+type ExitFarmResultType<BigUint> =
+    MultiValue2<EsdtTokenPayment<BigUint>, EsdtTokenPayment<BigUint>>;
+
+#[elrond_wasm::contract]
+pub trait Farm:
+    rewards::RewardsModule
+    + config::ConfigModule
+    + token_send::TokenSendModule
+    + locking_module::LockingModule
+    + farm_token::FarmTokenModule
+    + token_merge_helper::TokenMergeHelperModule
+    + farm_token_merge::FarmTokenMergeModule
+    + utils::UtilsModule
+    + pausable::PausableModule
+    + permissions_module::PermissionsModule
+    + sc_whitelist_module::SCWhitelistModule
+    + events::EventsModule
+    + elrond_wasm_modules::default_issue_callbacks::DefaultIssueCallbacksModule
+    + farm::base_functions::BaseFunctionsModule
+    + farm::exit_penalty::ExitPenaltyModule
+    + farm_base_impl::base_farm_init::BaseFarmInitModule
+    + farm_base_impl::base_farm_validation::BaseFarmValidationModule
+    + farm_base_impl::partial_positions::PartialPositionsModule
+    + farm_base_impl::enter_farm::BaseEnterFarmModule
+    + farm_base_impl::claim_rewards::BaseClaimRewardsModule
+    + farm_base_impl::compound_rewards::BaseCompoundRewardsModule
+    + farm_base_impl::exit_farm::BaseExitFarmModule
+    // farm boosted yields
+    + farm_boosted_yields::FarmBoostedYieldsModule
+    + week_timekeeping::WeekTimekeepingModule
+    + weekly_rewards_splitting::WeeklyRewardsSplittingModule
+    + weekly_rewards_splitting::ongoing_operation::OngoingOperationModule
+    + energy_query::EnergyQueryModule 
+{
+    #[init]
+    fn init(
+        &self,
+        reward_token_id: TokenIdentifier,
+        farming_token_id: TokenIdentifier,
+        division_safety_constant: BigUint,
+        pair_contract_address: ManagedAddress,
+        owner: ManagedAddress,
+        admins: MultiValueEncoded<ManagedAddress>,
+    ) {
+        self.base_farm_init(
+            reward_token_id,
+            farming_token_id,
+            division_safety_constant,
+            owner,
+            admins,
+        );
+
+        self.penalty_percent().set_if_empty(DEFAULT_PENALTY_PERCENT);
+        self.minimum_farming_epochs()
+            .set_if_empty(DEFAULT_MINUMUM_FARMING_EPOCHS);
+        self.burn_gas_limit().set_if_empty(DEFAULT_BURN_GAS_LIMIT);
+        self.pair_contract_address().set(&pair_contract_address);
+    }
+
+    #[payable("*")]
+    #[endpoint(enterFarm)]
+    fn enter_farm_endpoint(&self) -> EnterFarmResultType<Self::Api> {
+        let output_farm_token_payment = self.enter_farm();
```

Comment:
Shouldn't `enter_farm` also have a `caller` argument?

Reply:
For the enter farm algorithm we do not need the caller in order to compute the farming position, as the actual send of tokens towards the user does not happen in the base function. 
We will probably need the caller as an argument when the events module will be implemented, but it isn't in this PR.